File: f/find-kth-bit-in-nth-binary-string.cc
class Solution {
public:
  char findKthBit(int n, int k) {
    if (n == 1) return '0';

    int sz = 1 << n, h = sz / 2;
    if (k == h) return '1';
    else if (k < h) return findKthBit(n - 1, k);
    else return findKthBit(n - 1, sz - k) == '0' ? '1' : '0';
  }
};

File: f/find-kth-bit-in-nth-binary-string.erl
-spec find_kth_bit(N :: integer(), K :: integer()) -> char().
find_kth_bit(N, K) ->
  case kth_bit(N, K) of
    0 -> $0;
    1 -> $1
  end.

kth_bit(1, _K) ->
  0;
kth_bit(N, K) ->
  Mid = 1 bsl (N - 1),
  if
    K =:= Mid ->
      1;
    K < Mid ->
      kth_bit(N - 1, K);
    true ->
      1 - kth_bit(N - 1, (1 bsl N) - K)
  end.
